Add basic validation for route TLS configuration - checks that input is "syntactically" valid.

ramr opened the pull request:
Partial Fix for https://bugzilla.redhat.com/show_bug.cgi?id=1312292  

@smarterclayton PTAL Thx

The other PR https://github.com/openshift/origin/pull/8353 would be needed in addition to this for a complete fix but this one provides a partial fix to check input is valid PEM format.

```  
# make release  || echo "run make, copy binary and build new docker image"  
oadm router --latest-images  --replicas=0  
oc env dc/router EXTENDED_VALIDATION=true  
oc scale  dc/router --replicas=1  
```  
*Updated*  with testing instructions/usage.

smarterclayton commented on pkg/route/api/validation/validation.go:
```diff
@@ -11,9 +13,19 @@ import (
 )
 
+// disableRouteTLSValidation returns whether or not to disable TLS config validation checks.
+func disableRouteTLSValidation() bool {
```
Why not just create a new function and invoke it from the router in the unique_host check (or in an optional pre check step, based on a flag)?

ramr commented:
[test]

smarterclayton commented:
Can we check the cert order by validation?  We already do that in other
places in the code.

On Tue, Apr 5, 2016 at 10:19 PM, Ram Ranganathan <notifications@github.com>
wrote:

> Partial Fix for https://bugzilla.redhat.com/show_bug.cgi?id=1312292
>
> @smarterclayton <https://github.com/smarterclayton> PTAL Thx
>
> The other PR #8353 <https://github.com/openshift/origin/pull/8353> would
> be needed in addition to this for a complete fix but this one provides a
> partial fix to check input is valid PEM format.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/openshift/origin/pull/8366
> Commit Summary
>
>    - Add basic validation for route TLS configuration - checks that input
>    is
>
> File Changes
>
>    - *M* pkg/route/api/validation/validation.go
>    <https://github.com/openshift/origin/pull/8366/files#diff-0> (77)
>    - *M* pkg/route/api/validation/validation_test.go
>    <https://github.com/openshift/origin/pull/8366/files#diff-1> (257)
>
> Patch Links:
>
>    - https://github.com/openshift/origin/pull/8366.patch
>    - https://github.com/openshift/origin/pull/8366.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8366>
>

smarterclayton commented:
We don't have to validate in the api server - for instance, rejecting a route because the cert is malformed from the router side is sufficient.  My concern is letting haproxy do it (or having to do it for every single route) if the check is simple enough for us to do.

smarterclayton commented:
We can be more strict than what haproxy allows (especially if we give a good error message), as long as we can be sure we're not *less* strict.

ramr commented:
Ok - will rework this and get some bits in from the other PR.

ramr commented:
Ok @smarterclayton  PTAL  I pushed some changes. I added a check for cert verification + check key mismatches. Don't think we can do a full blown verification on cacerts or destcacerts (aka only support well known certifying authorities) as I suspect there could well be self signed ones on the edge [my testing] + private CAs on the haproxy-to-pod traffic end (re-encrypt) - i think we should allow for that.

ramr commented:
[test]

ramr commented:
[test]

smarterclayton commented:
Define full blown verification of cacerts?  What problems would openssl
have besides ordering?   Does openssl have a check function we could call?
I'm asking for an enumeration of possible failure modes, whether
verification checks them, and unknowns in haproxy.  If we don't know that,
we don't know whether the validation approach works.

On Thu, Apr 7, 2016 at 6:25 AM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/2779/)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8366#issuecomment-206800680>
>

ramr commented:
So what verification involves is - pem is valid, is a cert, checking the cert is valid (expiry + allowed domains), the certificate chain is valid (which would return an unknown authority for self-signed/generated CAs) and then optionally any extended key usages. We can do that for the edge  certificates we have as we have the associated ca cert but not for the cacert/destination cacert themselves. imho, this seems a bit backward to do in golang code anyway when `haproxy -c` would just verify this for us - which is what the PR https://github.com/openshift/origin/pull/8353 does. We are basically reverse engineering that and we cannot do any future-proofing guarantees anyway as haproxy code could change.

smarterclayton commented:
Can we test the certs in isolation rather than have to generate them into
the file?  I.e. construct a simple haproxy -c <simpleconfigfile> that
verifies only those certs?

On Thu, Apr 7, 2016 at 2:52 PM, Ram Ranganathan <notifications@github.com>
wrote:

> So what verification involves is - pem is valid, is a cert, checking the
> cert is valid (expiry + allowed domains), the certificate chain is valid
> (which would return an unknown authority for self-signed/generated CAs) and
> then optionally any extended key usages. We can do that for the edge
> certificates we have as we have the associated ca cert but not for the
> cacert/destination cacert themselves. imho, this seems a bit backward to do
> in golang code anyway when haproxy -c would just verify this for us -
> which is what the PR #8353 <https://github.com/openshift/origin/pull/8353>
> does. We are basically reverse engineering that and we cannot do any
> future-proofing guarantees anyway as haproxy code could change.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8366#issuecomment-207048143>
>

ramr commented:
So that was my original intent but then there was a point made regarding testing in isolation vs with the rest of the config as there could be potential of failures when run in "tandem". Which was the reason for doing the validate-config  on the generated config (just the new certs would be written but all the existing config + the config for the new service alias and its config/certs would be tested).

smarterclayton commented:
What would the tandem failure be?

ramr commented:
Due to overlapping backend configs. Some error cases I can think off could be due to with content switching overlapping frontend/backend modes, duplicate keys  (which we catch in the service aliases map), duplicates ids and acls (would have to be a custom config though).  Not a complete set by any stretch but just from a more defensive standpoint - checking in tandem vs isolation would address that and possibly be a bit more future-proof as it does catch more error cases.

smarterclayton commented:
The first couple you mentioned seem like things we have to catch for almost
all frameworks anyway - things under our control.

My concerns with #8353 (at least, as of now) are as follows:

* We have to check one at a time.  That's going to break any high density
case on restart.
* We have to change the structure of our queue to where it isn't layered
anymore (we're basically doing transactional commits, so each change
becomes an ordered transaction)
* Any break that is a user break needs to be communicated back to the user
somehow eventually


On Thu, Apr 7, 2016 at 4:00 PM, Ram Ranganathan <notifications@github.com>
wrote:

> Due to overlapping backend configs. Some error cases I can think off could
> be due to with content switching overlapping frontend/backend modes,
> duplicate keys (which we catch in the service aliases map), duplicates ids
> and acls (would have to be a custom config though). Not a complete set by
> any stretch but just from a more defensive standpoint - checking in tandem
> vs isolation would address that and possibly be a bit more future-proof as
> it does catch more error cases.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8366#issuecomment-207067063>
>

ramr commented:
So as re: the concerns on #8353 :
 1.  We can be smarter on that. A flag on the route - check if its set, skip otherwise. Set the flag
     if the route is modified (and set on creation).
2. We don't need a structure change to the queue really because we just apply the transactions in order. They are batched up, yes but the checks are always in order. If something fails a check, its not added to the batch. 
3. That we can do with updating the status on the route (which is similar to a duplicate route hostname).

smarterclayton commented:
Let's talk tomorrow.

ramr commented:
sounds good.

ramr commented:
[test]

smarterclayton commented on pkg/router/template/plugin.go:
```diff
@@ -42,6 +42,7 @@ type TemplatePluginConfig struct {
 	StatsUsername          string
 	StatsPassword          string
 	IncludeUDP             bool
+	ValidateRouteTLSConfig bool
```
Maybe this should be `ExtendedValidation` which more stringently tests the route against a set of known good safe configuration.

smarterclayton commented on pkg/router/template/router.go:
```diff
@@ -450,11 +458,38 @@ func (r *templateRouter) AddRoute(id string, route *routeapi.Route, host string)
 	}
 
+	var err error
+	if r.validateRouteTLSConfig {
```
Is this specific to template router?  What happens for F5?

ramr replied:
Yeah, its specific to the template router. F5 reports problems and ignores the config if a cert/key/ cacert etc is bad - does this both at admission time (config) or when it gets used (at run time).

smarterclayton commented on pkg/router/template/router.go:
```diff
@@ -450,11 +458,38 @@ func (r *templateRouter) AddRoute(id string, route *routeapi.Route, host string)
+		// Don't reprocess same errorneous config.
+		old, ok := frontend.InvalidServiceAliasConfigs[backendKey]
+		if ok && reflect.DeepEqual(old, config) {
+			// FIXME: Sending an error back causes an update
```
Isn't this fixed if you reject with a recorded error?

smarterclayton replied:
Or do you mean, this keeps checking the route in this code.

smarterclayton commented on pkg/router/template/router.go:
```diff
@@ -450,11 +458,38 @@ func (r *templateRouter) AddRoute(id string, route *routeapi.Route, host string)
 		}
 	}
 
+	var err error
```
Why not implement this whole chunk as an admission layer?  Extended validations, performs checks, keeps a map of checked routes (resourceVersion+uid)

smarterclayton replied:
(like unique hosts and the status update).  I ask because this structure makes it harder to refactor the router to ingress - while the admission approach means that by the time the template router gets the code it's known "good".

ramr replied:
Hmm, ok - that would be require rework - will look into it when am back on Tuesday then.

smarterclayton commented on pkg/route/api/validation/validation.go:
```diff
@@ -76,6 +79,64 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
 	return allErrs
 }
 
+// ValidateRouteTLSConfig checks if the TLS config (certificates and keys) for a route is valid.
```
@liggitt can you look over this chunk

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -76,6 +79,64 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
 }
 
+// ValidateRouteTLSConfig checks if the TLS config (certificates and keys) for a route is valid.
+func ValidateRouteTLSConfig(route *routeapi.Route) field.ErrorList {
```
all validation functions should take a fieldPath and build field paths for errors off of that parent path

ramr replied:
This is actually called from outside - i will call it extended route validation or something to that effect for clarity.

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -179,3 +240,71 @@ func validateInsecureEdgeTerminationPolicy(tls *routeapi.TLSConfig, fldPath *fie
+
+// validatePEMData checks if the PEM data is valid.
+func validatePEMData(pemData string) (*pem.Block, error) {
+	block, rest := pem.Decode([]byte(pemData))
```
there can be multiple PEM blocks in the cert, you can't assume the first one is the right one, or the only one. could this be a cert chain (multiple cert blocks)

ramr replied:
this was just validating the data format is valid - PEM. So checking that the block is not nil and rest is valid should suffice here.

ramr replied:
Never mind, I see what you mean - shows that am back from a break!! :^(

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -179,3 +240,71 @@ func validateInsecureEdgeTerminationPolicy(tls *routeapi.TLSConfig, fldPath *fie
+
+// validatePrivateKeyPEM checks if a private key PEM is valid.
+func validatePrivateKeyPEM(keyPEM string) error {
+	if _, err := validatePEMData(keyPEM); err != nil {
```
this doesn't validate anything about the key, just that the data contains a PEM block

ramr replied:
For now, yes - don't have complete info on how to check RSA/DSA/ECC keys. I'll put in a todo block here.

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -179,3 +240,71 @@ func validateInsecureEdgeTerminationPolicy(tls *routeapi.TLSConfig, fldPath *fie
+	}
+
+	if len(certPEM) > 0 {
+		if _, err := tls.X509KeyPair([]byte(certPEM), []byte(keyPEM)); err != nil {
```
validating the key+cert is odd inside a `validatePrivateKeyForCert` func... not really sure why we need both `validatePrivateKeyForCert` and `validatePrivateKeyPEM`

ramr replied:
Its to check that the combo is valid - private key and cert. It breaks at runtime/reload if the key is different. The `validatePrivateKeyPEM` is anyway just a stub for now.

liggitt replied:
> It breaks at runtime/reload if the key is different.

isn't that what we're trying to prevent?

ramr replied:
Yes, so that's what the `tls.X509KeyPair` checks. I am currently reworking this PR to simplify it - so hopefully the next push[es] tonight/tomorrow would make this cleaner and simpler.

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -76,6 +79,64 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
+	certPool := x509.NewCertPool()
+
+	if len(tls.CACertificate) > 0 {
+		cacert, err := validateCertificatePEM(tls.CACertificate, nil)
```
the CA cert especially could contain multiple cert PEM blocks... I'd expect the resulting pool to contain them all

ramr replied:
Why would we want to add them to the pool? We just want to check that the CA Certificate given to us is valid and the chain is valid and the leaf is signed by a well-known authority.

liggitt replied:
this is assuming there's only a single PEM block in the CACertificate `[]byte`... there could be multiple, and the pool we use to validate the cert needs to contain them all

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -76,6 +79,64 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
+	//       break, so disable the hostname validation for now.
+	// hostname := route.Spec.Host
+	hostname := ""
+	certPool := x509.NewCertPool()
```
leave as `nil` unless `len(tls.CACertificate) > 0`, so that verification uses system roots unless a CA is specified (which is what go does)

liggitt replied:
hmm... though I'm not sure what the router would do...

ramr replied:
Yeah, I wasn't sure on that as it depends on the configuration inside the router container but I'll make the change here to leave it nil - use the system roots.

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -76,6 +79,64 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
+				result = append(result, field.Invalid(tlsFieldPath.Child("key"), tls.Key, err.Error()))
+			}
+		} else {
+			if err := validatePrivateKeyPEM(tls.Key); err != nil {
```
when is a key without a certificate valid, regardless of whether it contains a PEM block? I'd expect an unconditional error here

ramr replied:
Well, today its really possible to specify both in one string - multiple PEM blocks in `tls.Certificate` and it would still be valid since the resulting PEM file contains required bits of information.

liggitt replied:
that still wouldn't make a Key without a Certificate make sense

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -76,6 +79,64 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
+	}
+
+	if len(tls.DestinationCACertificate) > 0 {
+		if _, err := validateCertificatePEM(tls.DestinationCACertificate, nil); err != nil {
```
same comment about multiple PEM blocks in the CA bundle

ramr replied:
same comment as above - this was just validating the data format is valid for all the PEM blocks.

liggitt replied:
DestinationCACertificate can contain multiple PEM blocks

ramr commented:
@smarterclayton  @liggitt  incorporated your feedback and made the changes and rebased. This is ready for your review.  Thx
Still pending an unit test for the extended_validator controller - will add that in tomorrow.

ramr commented:
[test]

ramr commented:
[test]

knobunc commented:
[test]

knobunc commented:
[test]

liggitt commented on pkg/cmd/infra/router/template.go:
```diff
@@ -192,7 +195,11 @@ func (o *TemplateRouterOptions) Run() error {
+	if o.ExtendedValidation {
+		nextPlugin = controller.NewExtendedValidator(nextPlugin, statusPlugin)
+	}
+	plugin := controller.NewUniqueHost(nextPlugin, o.RouteSelectionFunc(), statusPlugin)
```
cast `statusPlugin` to RejectionRecorder here so we know we're not using its plugin functions

ramr replied:
done.

liggitt commented on pkg/route/api/validation/validation.go:
```diff
@@ -76,6 +79,60 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
+		}
+	}
+
+	if len(tlsConfig.Key) > 0 {
```
do we support providing both the key and cert PEM blocks in the tls.Certificate field and leaving tls.Key empty? if so, that approach would skip this check

liggitt replied:
I'm not sure how we're combining/handling those cases in the router. Are we just concatenating the cert and key bytes and handing that off to the router? If so, I'd expect us to validate similarly here. Maybe something like:
```
if len(tlsConfig.Certificate) > 0 {
  certKeyBytes := []byte{}
  certKeyBytes = append(certKeyBytes, []byte(tlsConfig.Certificate)...)
  certKeyBytes = append(certKeyBytes, byte('\n'))
  certKeyBytes = append(certKeyBytes, []byte(tlsConfig.Key)...)
  if _, err := tls.X509KeyPair(certKeyBytes, certKeyBytes); err != nil {
    result = append(result, field.Invalid(tlsFieldPath.Child("key"), tlsConfig.Key, err.Error()))
  }
}
```

liggitt replied:
that all assumes, of course, that the go TLS PEM block identification is identical to the router's PEM block identification. I also don't know how the router reacts to multiple cert blocks (go concats all of them) or multiple key blocks (go takes the first one), or unidentified PEM blocks (go ignores them)

ramr replied:
Yeah, we just concatenate the certificate and key pem data to a file and then use that in the haproxy router. Since haproxy is reading this (and other certs) from a directory, it would present the first certificate for that host + I think its uses the first key. 
That's a good idea to just concatenate the cert and key data and pass it as both the key and cert pem files. Thx

smarterclayton replied:
Can we enforce those same behaviors at validation / transformation time?

ramr commented:
@liggitt made the fixes as per your recommendations, cleaned up remnants of the old implementation and added more tests.   @smarterclayton @knobunc  fyi.

knobunc commented:
For Online Beta they are disabling custom hostnames, so that means that we don't need to allow custom certs.  We need to work out how ops can disable certs (perhaps by deploying a custom router).

We also need to get this patch landed for GA.

ramr commented:
@smarterclayton /bump  anything else needed on this?

smarterclayton commented:
Regarding sni lookup order, is a crt-list the solution?  https://rafpe.ninja/2016/04/24/haproxy-ssl-domains-in-crt-list/

smarterclayton commented:
Will do final follow up on Monday.

liggitt commented:
> is a crt-list the solution?

Looks like it

smarterclayton commented:
[test]

smarterclayton commented:
[test]

smarterclayton commented:
Test failure is real

spinolacastro commented:
Do you believe it will land in time to v1.2.0 ?

smarterclayton commented:
No, this is unfortunately too big for 1.2.0.  You however will be able to use a 1.3.0.alpha track router image for that.

ramr commented:
rebased and fixed tests

ramr commented:
[test]

ramr commented:
@smarterclayton  PTAL - rebased and fixed tests + all tests pass now.

smarterclayton commented:
LGTM [merge]
